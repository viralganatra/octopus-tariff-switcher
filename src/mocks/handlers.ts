import { http, HttpResponse, graphql } from 'msw';
import {
  accountFixture,
  productAgileFixture,
  productCosyFixture,
  productGoFixture,
  productsFixture,
  standingChargeFixture,
  telemetryFixture,
  unitRatesFixture,
} from './fixtures';

export const handlers = [
  http.get('https://api.octopus.energy/v1/products', () => {
    return HttpResponse.json(productsFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/AGILE-24-10-01/', () => {
    return HttpResponse.json(productAgileFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/COSY-22-12-08/', () => {
    return HttpResponse.json(productCosyFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
    return HttpResponse.json(productGoFixture);
  }),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standing-charges/',
    () => {
      return HttpResponse.json(standingChargeFixture);
    },
  ),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standard-unit-rates',
    () => {
      return HttpResponse.json(unitRatesFixture);
    },
  ),
  graphql.mutation('ObtainKrakenToken', () => {
    return HttpResponse.json({
      data: {
        obtainKrakenToken: {
          token: 'foo',
        },
      },
    });
  }),
  graphql.query('Account', () => {
    return HttpResponse.json({
      data: accountFixture,
    });
  }),
  graphql.query('smartMeterTelemetry', () => {
    return HttpResponse.json({
      data: telemetryFixture,
    });
  }),
];
