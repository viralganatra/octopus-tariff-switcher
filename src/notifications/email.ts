import { z } from 'zod';
import { Resource } from 'sst';
import { API_MJML } from '../constants/api';
import { sendData } from '../utils/fetch';
import {
  getAlreadyCheapestTariffTemplate,
  getCheaperTariffTemplate,
  notWorthSwitchingTariffTemplate,
} from '../utils/email-template';
import { sendSparkPostEmail } from './sparkpost';
import { logger } from '../utils/logger';
import type { TariffContextWithCost } from '../types/tariff';
import type { EmailType } from '../types/email';

type SendEmail = {
  allTariffsByCost: TariffContextWithCost[];
  currentTariffWithCost: TariffContextWithCost;
  emailType: EmailType;
};

async function getHtml(content: string) {
  const mjmlAuth = `${Resource.MjmlAppId.value}:${Resource.MjmlSecretKey.value}`;

  logger.info(`API: Getting email template data via ${API_MJML}`);

  const schema = z.object({
    html: z.string(),
  });

  const results = await sendData({
    url: API_MJML,
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${btoa(mjmlAuth)}`,
    },
    body: { mjml: content },
  });

  const { html } = schema.parse(results);

  logger.info('API Response: Received email template data');

  return html;
}

export async function sendEmail({ allTariffsByCost, currentTariffWithCost, emailType }: SendEmail) {
  let emailSubject: string;
  let emailContent: string;

  switch (emailType) {
    case 'CHEAPER_TARIFF_EXISTS':
      emailSubject = 'Cheaper Tariff Found';
      emailContent = getCheaperTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
    case 'ALREADY_ON_CHEAPEST_TARIFF':
      emailSubject = 'You are already on the cheapest tariff';
      emailContent = getAlreadyCheapestTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
    case 'NOT_WORTH_SWITCHING_TARIFF':
      emailSubject = 'Not worth switching tariff';
      emailContent = notWorthSwitchingTariffTemplate({
        allTariffsByCost,
        currentTariffWithCost,
        title: emailSubject,
      });
      break;
  }

  const html = await getHtml(emailContent);

  return sendSparkPostEmail({ html, subject: emailSubject });
}
