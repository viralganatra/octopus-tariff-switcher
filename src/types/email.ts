import type { TariffContextWithCost } from './tariff';

export type EmailType =
  | 'CHEAPER_TARIFF_EXISTS'
  | 'ALREADY_ON_CHEAPEST_TARIFF'
  | 'NOT_WORTH_SWITCHING_TARIFF';

export type SendEmail = {
  allTariffsByCost: TariffContextWithCost[];
  currentTariffWithCost: TariffContextWithCost;
  emailType: EmailType;
};
