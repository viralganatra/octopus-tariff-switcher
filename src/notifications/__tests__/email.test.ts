import { sendEmail } from '../email';
import { server } from '../../mocks/node';

describe('Email', () => {
  it('call the email render api with the correct data', async () => {
    const dispatchRequest = vi.fn();

    server.events.on('request:start', dispatchRequest);

    await sendEmail({
      emailType: 'CHEAPER_TARIFF_EXISTS',
      allTariffsByCost: [],
      currentTariffWithCost: {
        id: 'agile',
        displayName: 'Agile Octopus',
        tariffCodeMatcher: '-AGILE-',
        costInPence: 100,
      },
    });

    // @ts-ignore
    const serverRequest = dispatchRequest.mock.calls[0][0].request;

    expect(serverRequest.url).toBe('https://api.mjml.io/v1/render');
    expect(serverRequest.headers.get('content-type')).toBe('application/x-www-form-urlencoded');
    expect(serverRequest.headers.get('authorization')).toBe(
      'Basic TWptbEFwcElkOk1qbWxTZWNyZXRLZXk=',
    );
  });

  it('should call the sparkpost api with the correct data', async () => {
    const result = (await sendEmail({
      emailType: 'CHEAPER_TARIFF_EXISTS',
      allTariffsByCost: [],
      currentTariffWithCost: {
        id: 'agile',
        displayName: 'Agile Octopus',
        tariffCodeMatcher: '-AGILE-',
        costInPence: 100,
      },
    })) as { url: string; headers: { Authorization: string } };

    expect(result.url).toBe('https://api.eu.sparkpost.com/api/v1/transmissions');
    expect(result.headers.Authorization).toBe('SparkPostApiKey');
  });

  it('should send an email when a cheaper tariff exists', async () => {
    const result = (await sendEmail({
      emailType: 'CHEAPER_TARIFF_EXISTS',
      allTariffsByCost: [
        {
          costInPence: 95,
          displayName: 'Octopus Cosy',
          id: 'cosy',
          tariffCodeMatcher: '-COSY-',
        },
        {
          id: 'agile',
          displayName: 'Agile Octopus',
          tariffCodeMatcher: '-AGILE-',
          costInPence: 105,
        },
      ],
      currentTariffWithCost: {
        id: 'agile',
        displayName: 'Agile Octopus',
        tariffCodeMatcher: '-AGILE-',
        costInPence: 105,
      },
    })) as { data: object };

    expect(result.data).toMatchSnapshot();
  });

  it('should send an email when there is no cheaper tariff', async () => {
    const result = (await sendEmail({
      emailType: 'ALREADY_ON_CHEAPEST_TARIFF',
      allTariffsByCost: [
        {
          costInPence: 105,
          displayName: 'Octopus Cosy',
          id: 'cosy',
          tariffCodeMatcher: '-COSY-',
        },
        {
          id: 'agile',
          displayName: 'Agile Octopus',
          tariffCodeMatcher: '-AGILE-',
          costInPence: 65,
        },
      ],
      currentTariffWithCost: {
        id: 'agile',
        displayName: 'Agile Octopus',
        tariffCodeMatcher: '-AGILE-',
        costInPence: 65,
      },
    })) as { data: object };

    expect(result.data).toMatchSnapshot();
  });

  it('should send an email when it is not worth switching tariff', async () => {
    const result = (await sendEmail({
      emailType: 'NOT_WORTH_SWITCHING_TARIFF',
      allTariffsByCost: [
        {
          costInPence: 99,
          displayName: 'Octopus Cosy',
          id: 'cosy',
          tariffCodeMatcher: '-COSY-',
        },
        {
          id: 'agile',
          displayName: 'Agile Octopus',
          tariffCodeMatcher: '-AGILE-',
          costInPence: 100,
        },
      ],
      currentTariffWithCost: {
        id: 'agile',
        displayName: 'Agile Octopus',
        tariffCodeMatcher: '-AGILE-',
        costInPence: 100,
      },
    })) as { data: object };

    expect(result.data).toMatchSnapshot();
  });
});
