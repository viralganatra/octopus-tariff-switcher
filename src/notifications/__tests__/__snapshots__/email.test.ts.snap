// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Email > should send an email when a cheaper tariff exists 1`] = `
{
  "content": {
    "from": {
      "email": "EmailFrom",
      "name": "Octobot",
    },
    "html": "
    <mjml>
      <mj-body background-color="white">
        <mj-section background-color="#f0f0f0">
          <mj-column width="15%" vertical-align="middle">
            <mj-image src="https://i.postimg.cc/v4fmRR64/businessman.png"></mj-image>
          </mj-column>
          <mj-column width="85%" vertical-align="middle">
        <mj-text font-style="bold" font-size="24px" color="#525252" font-family="Helvetica Neue">Octopus Tariff Switcher</mj-text>
       </mj-column>
      </mj-section>
        
    <mj-section>
      <mj-column width="60%">
        <mj-text font-style="bold" font-size="24px" line-height="30px" font-family="Helvetica Neue" color="#525252">Cheaper Tariff Found.</mj-text>
          <mj-table font-size="16px" line-height="24px" font-family="Helvetica Neue" color="#525252">
            <tr style="border-bottom:1px solid #ecedee;text-align:left;padding:15px 0;">
              <th style="padding: 8px">Tariff</th>
              <th style="padding: 8px">Cost</th>
            </tr>
            
                <tr style="background-color:#f0f0f0">
                  <td style="padding: 8px">Octopus Cosy</td>
                  <td style="padding: 8px">£0.95</td>
                </tr>
              
                <tr >
                  <td style="padding: 8px">Agile Octopus</td>
                  <td style="padding: 8px">£1.05</td>
                </tr>
              
          </mj-table>
          
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on Agile Octopus. I have automatically switched to Octopus Cosy as it is cheaper.
      </mj-text>
      <mj-text font-size="16px" line-height="24px" padding-top="10px" color="#525252">
        You will pay £0.95 versus £1.05, saving £0.10.
      </mj-text>
    
        </mj-column>
        <mj-column width="40%">
          <mj-image src="https://i.postimg.cc/y8DtP8TQ/robot-cheaper-tariff.png"></mj-image>
        </mj-column>
      </mj-section>
  
      </mj-body>
    </mjml>
  ",
    "subject": "Octopus Tariff Switcher: Cheaper Tariff Found",
  },
  "options": {
    "transactional": true,
  },
  "recipients": {
    "list_id": "octopus-tariff-switcher",
  },
}
`;

exports[`Email > should send an email when it is not worth switching tariff 1`] = `
{
  "content": {
    "from": {
      "email": "EmailFrom",
      "name": "Octobot",
    },
    "html": "
    <mjml>
      <mj-body background-color="white">
        <mj-section background-color="#f0f0f0">
          <mj-column width="15%" vertical-align="middle">
            <mj-image src="https://i.postimg.cc/v4fmRR64/businessman.png"></mj-image>
          </mj-column>
          <mj-column width="85%" vertical-align="middle">
        <mj-text font-style="bold" font-size="24px" color="#525252" font-family="Helvetica Neue">Octopus Tariff Switcher</mj-text>
       </mj-column>
      </mj-section>
        
    <mj-section>
      <mj-column width="60%">
        <mj-text font-style="bold" font-size="24px" line-height="30px" font-family="Helvetica Neue" color="#525252">Not worth switching tariff.</mj-text>
          <mj-table font-size="16px" line-height="24px" font-family="Helvetica Neue" color="#525252">
            <tr style="border-bottom:1px solid #ecedee;text-align:left;padding:15px 0;">
              <th style="padding: 8px">Tariff</th>
              <th style="padding: 8px">Cost</th>
            </tr>
            
                <tr style="background-color:#f0f0f0">
                  <td style="padding: 8px">Octopus Cosy</td>
                  <td style="padding: 8px">£0.99</td>
                </tr>
              
                <tr >
                  <td style="padding: 8px">Agile Octopus</td>
                  <td style="padding: 8px">£1.00</td>
                </tr>
              
          </mj-table>
          
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on Agile Octopus. It isn't worth switching to the other tariffs as the saving is too little.
      </mj-text>
    
        </mj-column>
        <mj-column width="40%">
          <mj-image src="https://i.postimg.cc/JzSgdC61/robot-not-worth-sw-itching.png"></mj-image>
        </mj-column>
      </mj-section>
  
      </mj-body>
    </mjml>
  ",
    "subject": "Octopus Tariff Switcher: Not worth switching tariff",
  },
  "options": {
    "transactional": true,
  },
  "recipients": {
    "list_id": "octopus-tariff-switcher",
  },
}
`;

exports[`Email > should send an email when there is no cheaper tariff 1`] = `
{
  "content": {
    "from": {
      "email": "EmailFrom",
      "name": "Octobot",
    },
    "html": "
    <mjml>
      <mj-body background-color="white">
        <mj-section background-color="#f0f0f0">
          <mj-column width="15%" vertical-align="middle">
            <mj-image src="https://i.postimg.cc/v4fmRR64/businessman.png"></mj-image>
          </mj-column>
          <mj-column width="85%" vertical-align="middle">
        <mj-text font-style="bold" font-size="24px" color="#525252" font-family="Helvetica Neue">Octopus Tariff Switcher</mj-text>
       </mj-column>
      </mj-section>
        
    <mj-section>
      <mj-column width="60%">
        <mj-text font-style="bold" font-size="24px" line-height="30px" font-family="Helvetica Neue" color="#525252">You are already on the cheapest tariff.</mj-text>
          <mj-table font-size="16px" line-height="24px" font-family="Helvetica Neue" color="#525252">
            <tr style="border-bottom:1px solid #ecedee;text-align:left;padding:15px 0;">
              <th style="padding: 8px">Tariff</th>
              <th style="padding: 8px">Cost</th>
            </tr>
            
                <tr style="background-color:#f0f0f0">
                  <td style="padding: 8px">Octopus Cosy</td>
                  <td style="padding: 8px">£1.05</td>
                </tr>
              
                <tr >
                  <td style="padding: 8px">Agile Octopus</td>
                  <td style="padding: 8px">£0.65</td>
                </tr>
              
          </mj-table>
          
      <mj-text font-size="16px" line-height="24px" padding-top="20px" color="#525252">
        You are currently on Agile Octopus. It isn't worth switching as the other tariffs are more expensive.
      </mj-text>
    
        </mj-column>
        <mj-column width="40%">
          <mj-image src="https://i.postimg.cc/NfJZmHR2/robot-already-on-cheapest.png"></mj-image>
        </mj-column>
      </mj-section>
  
      </mj-body>
    </mjml>
  ",
    "subject": "Octopus Tariff Switcher: You are already on the cheapest tariff",
  },
  "options": {
    "transactional": true,
  },
  "recipients": {
    "list_id": "octopus-tariff-switcher",
  },
}
`;
